3d_segmentation/torch/compute_metric.py

```diff
@@ -0,0 +1,131 @@
+  Alternatively, we can also use `torch.multiprocessing.spawn` to start program, but it that case, need to handle
+  all the above parameters and compute `rank` manually, then set to `init_process_group`, etc.
+  `torch.distributed.launch` is even more efficient than `torch.multiprocessing.spawn`.
+- Use `init_process_group` to initialize every process, every GPU runs in a separate process with unique rank.
+  Here we use `NVIDIA NCCL` as the backend and must set `init_method="env://"` if use `torch.distributed.launch`.
+- Partition the saved predictions and labels into ranks for parallel computation.
+- Compute `Dice Metric` on every process, reduce the results after synchronization.
+
+Note:
+    `torch.distributed.launch` will launch `nnodes * nproc_per_node = world_size` processes in total.
+    Suggest setting exactly the same software environment for every node, especially `PyTorch`, `nccl`, etc.
+    A good practice is to use the same MONAI docker image for all nodes directly.
+    Example script to execute this program on every node:
+    python -m torch.distributed.launch --nproc_per_node=NUM_GPUS_PER_NODE
+           --nnodes=NUM_NODES --node_rank=INDEX_CURRENT_NODE
+           --master_addr="localhost" --master_port=1234
+           compute_metric.py
+
+    This example was tested with [Ubuntu 16.04/20.04], [NCCL 2.6.3].
+
+Referring to: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
+
+"""
+
+import argparse
+import os
+from glob import glob
+
+import nibabel as nib
+import numpy as np
+import torch
+import torch.distributed as dist
+from torch.nn.parallel import DistributedDataParallel
+
+import monai
+from monai.data import DataLoader, Dataset, partition_dataset
+from monai.handlers import write_metrics_reports
+from monai.inferers import sliding_window_inference
+from monai.metrics import DiceMetric
+from monai.transforms import EnsureChannelFirstd, Compose, LoadImaged, ToTensord
+from monai.utils import string_list_all_gather
+
+
+def compute(args):
+    # initialize the distributed evaluation process, every GPU runs in a process
+    dist.init_process_group(backend="nccl", init_method="env://")
+    device = torch.device(f"cuda:{args.local_rank}")
+    torch.cuda.set_device(device)
+
+    preds = sorted(glob(os.path.join(args.dir, "im*", "im*_seg.nii.gz")))
+    labels = sorted(glob(os.path.join(args.dir, "seg*.nii.gz")))
+    datalist = [{"pred": pred, "label": label} for pred, label in zip(preds, labels)]
+
+    # split data for every subprocess, for example, 16 processes compute in parallel
+    data_part = partition_dataset(
+        data=datalist,
+        num_partitions=dist.get_world_size(),
+        shuffle=False,
+        even_divisible=False,
+    )[dist.get_rank()]
+
+    # define transforms for predictions and labels
+    transforms = Compose(
+        [
+            LoadImaged(keys=["pred", "label"]),
+            EnsureChannelFirstd(keys=["pred", "label"]),
```

Comment:
looks good, could you make synthetic images in this demo to make it runnable? would be great to have some postprocessing transforms here such as `KeepLargestConnectedComponentD` it is often used during evaluation

Reply:
Hi @wyli ,

Thanks for your review.
This demo is runnable, it uses the saved `preds` and `labels` from the `unet_evaluation_dict.py` program, do you mean to copy an inference logic here?

Thanks.

Reply:
no, I mean having a simple and self-contained example for metrics computation

Reply:
So you mean to do below steps?
1. generate synthetic `images` and `labels`.
2. run evaluation with `images` and `labels`.(with `KeepLargestConnectedComponentD` as post transform)
3. save the `predictions` and `labels`.
4. compute metrics from saved `predictions` and `labels`.
5. generate metrics report.

Thanks.

Reply:
1. generate predictions and labels (with KeepLargest on predictions)
2. compute metrics
3. generate a report
this could be a new script in the modules/ folder

Reply:
OK, I see, let me investigate how to generate `preds` and `labels` directly, I thought our synthetic API can only generate `(image, label)` pair.

Thanks.

Reply:
updated in the latest commit.
Thanks.